### ISSUE TITLE
More dynamic display of menu element within dropdown component 

### ISSUE THREAD

**Issue by brenner-company** [appuniversum/ember-appuniversum#333]
**After replacing our own dropdown component with `AuDropdown` we encountered the following (non-breaking) issue:**
When using a dropdown within a container with a defined overflow property (`auto`, `scroll`, etc.) we noticed that the content of that container jumped position to display the menu when there is insufficient space. Also, when the trigger for that dropdown is for example at the bottom of a container (or the page in general) with its overflow hidden,  the menu can be clipped even if there is enough space on top of the trigger.

A example of the first issue can be found below:

![CleanShot 2022-10-19 at 11 57 26@2x](https://user-images.githubusercontent.com/18174827/196714060-62b86ff1-4fe7-4edc-9fc5-190d61db0e3a.png)

![CleanShot 2022-10-19 at 11 57 40@2x](https://user-images.githubusercontent.com/18174827/196720490-0819860d-9e5d-4c4f-8b9d-cc81d5ecb478.png)

**I think the cause of the issues here is:**
-  the fact that the menu is displayed alongside the trigger (inside the HTML) instead of at the end of the `body` element for example.
- the fact that the positioning the menu is done solely by preset and doesn't have functionality to calculate a back-up position if there is insufficient space.

These are two things what most external popover/tooltip plugins seem to provide.

Is this something that we could provide ourselves? Or could we use something like https://github.com/kybishop/ember-attacher (we used this within Kaleidos for our own dropdown component) to enable this?

**Comment by brenner-company** [comment]
I made a quick prove of concept (of dynamic positioning) using Popper JS (https://popper.js.org/ with https://github.com/NullVoxPopuli/ember-popperjs) and it seems to work well. Could this be an option?

See following file changes: https://github.com/appuniversum/ember-appuniversum/compare/feature/dynamic-display-of-menu-within-dropdown

Don't mind the bad JS and commented SCSS for the moment 😉

**Comment by brenner-company** [comment]
There also seems to be an option to attach the menu to the `body` element (this fixes the container jump issue when `auto`, `scroll`, etc. is defined), but I don't have any time left today to give it a look yet.

**Comment by Windvis** [comment]
I think using something like popper could be a good idea, yes 👍. It could probably be useful for other features in the future as well. The only problem at the moment is that `ember-popperjs` requires Ember v3.25+ and we still support v3.24 in Appuniversum v1. So we can't release it as part of v1 like that.

(I should just finish up the v2 work which bumps the support to v3.28).

I'm also a bit hesitant since popper has been superseded by [floating-ui](https://github.com/floating-ui/floating-ui/) so it feels like adding tech debt already 😄.

In any case, I think it shouldn't be to hard to create a custom modifier with just the basic options to unblock the dropdown issue you are seeing. And that could also still support Ember 3.24.

**Comment by brenner-company** [comment]
> In any case, I think it shouldn't be to hard to create a custom modifier with just the basic options to unblock the dropdown issue you are seeing. And that could also still support Ember 3.24.

By that, you mean a modifier for the floating-ui library, right? If that's the case, I can give it a go if I can base it of a good example (of something similar). I'm not that well acquainted with Ember yet 😊

**Comment by Windvis** [comment]
> By that, you mean a modifier for the floating-ui library, right?

Indeed. The Ember side isn't that different from the vanilla js implementation I think. The [`ember-modifier`](https://github.com/ember-modifier/ember-modifier) addon needs to be installed as a dependency but it's pretty useful. The docs can explain it better than I can 😄.

Modifiers give you access to the DOM element you attach it to, so no need for selectors. To render the dropdown content in the "root" element I think we can use a similar setup as the [modal component](https://github.com/appuniversum/ember-appuniversum/blob/59073eaa55b23f70720da7ab08a1c4d2e5f1c57c/addon/components/au-modal.hbs#L2). So the AuDropdown would find the wormhole element and pass it along to the modifier. Something like that.

Feel free to ping me here or in the Rocket.chat if things are unclear or have questions 👍.
